### PR TITLE
Add Custom Headers To Config Parser

### PR DESCRIPTION
**Pull request by DanielleMiu**
adds two additional configuration settings to the `listener` config section:
`custom_api_response_headers` and `custom_ui_response_headers`

this update also adds `WrapCustomHeadersHandler` wrapper which includes all of the logic necessary to perform custom header injection automatically

### CODE REVIEW

File: listenerutil/parse.go [hunk 1]
```diff
@@ -387,3 +413,122 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
...
+	"1xx",
+	"2xx",
+	"3xx",
+	"4xx",
+	"5xx",
+}
+
+const strictTransportSecurity = "max-age=31536000; includeSubDomains"
+
+// parseCustomResponseHeaders takes a raw config values for the
```
Comment (johanbrandhorst) [inline listenerutil/parse.go:428]:
```suggestion
// parseCustomResponseHeaders takes a raw config value for the
```

File: listenerutil/parse.go [hunk 2]
```diff
@@ -387,3 +413,122 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
...
+	h := make(map[string]map[string]string)
+	// if r is nil, we still should set the default custom headers
+	if responseHeaders == nil {
+		h["default"] = map[string]string{"Strict-Transport-Security": strictTransportSecurity}
+		return h, nil
+	}
+
+	customResponseHeader, ok := responseHeaders.([]map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("response headers were not configured correctly. please make sure they're in a slice of maps")
```
Comment (johanbrandhorst) [inline listenerutil/parse.go:443]: The context in which a user will see this error is if they misconfigure their HCL. We should adapt the error message. Per the [HCL definitions](https://github.com/hashicorp/hcl/blob/main/hclsyntax/spec.md), I think we should use "list".
```suggestion
		return nil, fmt.Errorf("response headers were not configured correctly. Please make sure they're in a list of maps")
```

File: listenerutil/parse.go [hunk 3]
```diff
@@ -387,3 +413,122 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
...
+		h["default"] = make(map[string]string)
+	}
+	if _, ok := h["default"]["Strict-Transport-Security"]; !ok {
+		h["default"]["Strict-Transport-Security"] = strictTransportSecurity
+	}
+
+	return h, nil
+}
+
+// isValidStatusCode checking for status codes outside the boundary
```
Comment (johanbrandhorst) [inline listenerutil/parse.go:481]: This could be a tiny bit confusing 😂.
```suggestion
// isValidStatusCode checks for status codes outside the allowed range
```

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]
I realise you probably meant to mark this a draft, but lets make sure we add tests before we merge this.

File: configutil/config_test.go [hunk 4]
```diff
@@ -45,6 +47,104 @@ func TestParseConfig(t *testing.T) {
...
+						"test" =          ["ui default value"]
+						"cache-control" = ["max-age=604800"]
+					}
+					"200" = {
+						"test" = ["ui 200 value"]
+					}
+					"401" = {
+						"test" = ["ui 401 value"]
+					}
+				}
```
Comment (johanbrandhorst) [inline configutil/config_test.go:76]: Could we add some `2xx`, `3xx` etc?

File: listenerutil/listener.go [hunk 5]
```diff
@@ -146,6 +146,7 @@ PASSPHRASECORRECT:
 				badCiphers = append(badCiphers, cipherStr)
 			}
 		}
+		// TODO: change this to not use a vault-specific error
```
Comment (johanbrandhorst) [inline listenerutil/listener.go:149]: Is this happening :P? Maybe in a separate PR?

File: listenerutil/parse.go [hunk 6]
```diff
@@ -387,3 +413,139 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
...
+	"3xx",
+	"4xx",
+	"5xx",
+}
+
+const strictTransportSecurity = "max-age=31536000; includeSubDomains"
+const xContentTypeOptions = "nosniff"
+const cacheControl = "no-store"
+const apiContentSecurityPolicy = "default-src 'none'"
+const uiContentSecurityPolicy = "default-src 'none'; script-src 'self'; frame-src 'self'; font-src 'self'; connect-src 'self'; img-src 'self' data:*; style-src 'self'; media-src 'self'; manifest-src 'self'; style-src-attr 'self'; frame-ancestors 'self'"
```
Comment (johanbrandhorst) [inline listenerutil/parse.go:430]: Is there a reference for why we chose this as the default CSP?

File: listenerutil/parse.go [hunk 7]
```diff
@@ -387,3 +413,139 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
...
+	}
+	return hvMap, nil
+}
+
+func parseHeaderValues(header interface{}) (string, error) {
+	var sl []string
+	if _, ok := header.([]interface{}); !ok {
+		return "", fmt.Errorf("headers must be given in a list of strings")
+	}
+	headerValList := header.([]interface{})
```
Comment (johanbrandhorst) [inline listenerutil/parse.go:536]:
```suggestion
	headerValList, ok := header.([]interface{})
	if !ok {
		return "", fmt.Errorf("headers must be given in a list of strings")
	}
```

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]

File: listenerutil/listener.go (hunk 5 above)
Comment (DanielleMiu) [reply listenerutil/listener.go:149]: this is mostly for future reference, though it is such a small change that I *could* add it in this pr?

Comment (johanbrandhorst) [reply listenerutil/listener.go:149]: Go for it IMO

File: listenerutil/parse.go (hunk 6 above)
Comment (DanielleMiu) [reply listenerutil/parse.go:430]: these are the headers that were defined in the ticket

Comment (johanbrandhorst) [reply listenerutil/parse.go:430]: I couldn't find a reference in the ticket so w/e I guess 😛

File: listenerutil/handler.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+		}
+	}
+
+	// Setting the default headers first
+	if val, ok := sch[0]; ok {
+		setter(val)
+	}
+
+	// Then setting the generic hundred-level headers
+	if val, ok := sch[statusCode/100]; ok {
```
Comment (johanbrandhorst) [inline listenerutil/handler.go:61]: Could you add a comment here noting that this uses integer division, which always rounds down?
```suggestion
	// Note: this integer division always rounds down, so 499/100 = 4
	if val, ok := sch[statusCode/100]; ok {
```

File: listenerutil/handler.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+		setter(val)
+	}
+
+	// Finally setting the status-specific headers
+	if val, ok := sch[statusCode]; ok {
+		setter(val)
+	}
```
Comment (johanbrandhorst) [inline listenerutil/handler.go:68]: What is the behavior if multiple different levels specify conflicting header values? Should we use `w.Header().Add` instead in the setter? What happens for duplicates?

File: listenerutil/handler.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+type uiRequestFunc func(*http.Request) bool
+
+// WrapCustomHeadersHandler wraps the handler to pass a custom ResponseWriter struct to all
+// later wrappers and handlers to assign custom headers by status code. This wrapper must
+// be the outermost wrapper to function correctly.
```
Comment (johanbrandhorst) [inline listenerutil/handler.go:75]: The terminology for this is a HTTP middleware, lets use that in the API docs. Also, this describes the internal implementations details, when as a user I just want to know that it does what I want with the config.
```suggestion
// WrapCustomHeadersHandler wraps the handler, adding custom headers as defined in the
// ListenerConfig. The isUiRequest callback is used to determine whether an incoming request
// should get the UI specific headers or the generic API headers. This middleware must
// be the outermost middleware to function correctly in the case of another middleware returning
// early.
```

Comment (johanbrandhorst) [reply listenerutil/handler.go:75]: Actually that reminds me, what happens if this middleware wraps a handler that returns without calling w.WriteHeader? Should we call w.WriteHeader for good measure before returning from here?

File: listenerutil/parse.go [hunk 11]
```diff
@@ -387,3 +413,153 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
...
+const uiContentSecurityPolicy = "default-src 'none'; script-src 'self'; frame-src 'self'; font-src 'self'; connect-src 'self'; img-src 'self' data:*; style-src 'self'; media-src 'self'; manifest-src 'self'; style-src-attr 'self'; frame-ancestors 'self'"
+
+// parseCustomResponseHeaders takes raw config values for the "custom_ui_response_headers"
+// and "custom_api_response_headers". It makes sure the config entry is passed in as a map
+// of status code to a map of header name and header values. It verifies the validity of the
+// status codes, and header values. It also adds the default headers values for "Cache-Control",
+// "Strict-Transport-Security", "X-Content-Type-Options", and "Content-Security-Policy".
+func parseCustomResponseHeaders(responseHeaders interface{}, uiHeaders bool) (map[int]map[string]string, error) {
+	h := make(map[int]map[string]string)
+	// if r is nil, we still should set the default custom headers
```
Comment (johanbrandhorst) [inline listenerutil/parse.go:439]:
```suggestion
	// if responseHeaders is nil, we still should set the default custom headers
```

File: listenerutil/parse.go [hunk 12]
```diff
@@ -387,3 +413,153 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
...
+			return "", fmt.Errorf("found a non-string header value: %v", vh)
+		}
+		headerVal := strings.TrimSpace(vh.(string))
+		if headerVal == "" {
+			continue
+		}
+		sl = append(sl, headerVal)
+
+	}
+	s := strings.Join(sl, "; ")
```
Comment (johanbrandhorst) [inline listenerutil/parse.go:562]: Instead of joining the strings like this, can we not keep the separate values as a slice and just add all of them with `Header().Add(v)`? This feels a bit ugly to be doing here.

File: listenerutil/handler.go (hunk 9 above)
Comment (DanielleMiu) [reply listenerutil/handler.go:68]: the latest call to `Header().Set()` overrides previous headers added the same way

File: listenerutil/handler.go (hunk 10 above)
Comment (DanielleMiu) [reply listenerutil/handler.go:75]: I have another comment addressing this in the code [here](https://github.com/hashicorp/go-secure-stdlib/pull/49/files#diff-edb1c099ae693e6b00aa24b159e9cd9cb6aca1bab17f5d168bf0ae5b90757138R28), but the default behavior of `ResponseWriter` is such that if a call has not been made to `WriteHeader` when `Write` is called, then `WriteHeader(200)` is called automatically

File: listenerutil/handler.go (hunk 9 above)
Comment (johanbrandhorst) [reply listenerutil/handler.go:69]: Is that the behavior we want? I could imagine saying you want both the 4xx and 401 headers to be merged into one.

File: listenerutil/handler.go (hunk 10 above)
Comment (johanbrandhorst) [reply listenerutil/handler.go:76]: Yes, but that will happen outside of our custom response writer, will it not?

File: listenerutil/parse.go (hunk 12 above)
Comment (johanbrandhorst) [reply listenerutil/parse.go:553]: I don't think either of those are good enough reasons to keep this TBH. We should operate on these as well typed values until we know we want to merge them, and at that point we shouldn't be doing it manually with `;`, we should use the library tools (`w.Header().Add`), IMO.

File: listenerutil/handler.go (hunk 9 above)
Comment (DanielleMiu) [reply listenerutil/handler.go:69]: handled offline, this is the behavior we want, since we need overrides to work

File: listenerutil/handler.go (hunk 10 above)
Comment (DanielleMiu) [reply listenerutil/handler.go:76]: if you check out the link I posted, you can see the logic is replicated in the response writer and is even tested for in the handler tests for the 200 responses

Comment (johanbrandhorst) [reply listenerutil/handler.go:79]: My concern is what happens if `Write` isn't called but the inner middleware simply returns? It'd get all the way out of the middleware chain and then call `WriteHeader` on a response writer that isn't ours, no?

File: configutil/go.mod [hunk 13]
```diff
@@ -2,6 +2,9 @@ module github.com/hashicorp/go-secure-stdlib/configutil/v2
 
 go 1.17
 
+// TODO: remove
+replace github.com/hashicorp/go-secure-stdlib/listenerutil => ../listenerutil
+
```
Comment (johanbrandhorst) [inline configutil/go.mod:7]: Lets not forget to remove this before merge 😁

File: listenerutil/handler.go [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+func WrapCustomHeadersHandler(h http.Handler, config *ListenerConfig, isUiRequest uiRequestFunc) http.Handler {
+	// TODO: maybe we should perform some preparsing here on the headers? check for duplicates,
+	// headers that aren't allowed, etc.
```
Comment (johanbrandhorst) [inline listenerutil/handler.go:82]: Do you want to remove this comment? I would expect this sort of parsing to happen in the parser, not here.

File: listenerutil/handler_test.go [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+
+			resp := w.Result()
+			assert.Equal(t, tt.expStatus, resp.StatusCode)
+
+			for header, expected := range tt.expHeaders {
+				assert.Empty(t, cmp.Diff(expected, resp.Header.Values(header)))
+			}
+		})
+	}
+}
```
Comment (johanbrandhorst) [inline listenerutil/handler_test.go:243]: Great tests!

**Review by johanbrandhorst — DISMISSED** [review]

File: listenerutil/parse.go [hunk 16]
```diff
@@ -387,3 +413,142 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
 		return "", fmt.Errorf("multiple addresses found (%q), please configure one", out)
 	}
 }
+
+const strictTransportSecurity = "max-age=31536000; includeSubDomains"
```
Comment (jefferai) [inline listenerutil/parse.go:417]: Can you name these with a bit more definition? e.g. `defaultStrictTransportSecurityHeader`, `defaultXContentTypeOptionsHeader", `defaultCacheControlHeader`, etc.

File: listenerutil/parse.go [hunk 17]
```diff
@@ -387,3 +413,142 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
...
+// and "custom_api_response_headers". It makes sure the config entry is passed in as a map
+// of status code to a map of header name and header values. It verifies the validity of the
+// status codes, and header values. It also adds the default headers values for "Cache-Control",
+// "Strict-Transport-Security", "X-Content-Type-Options", and "Content-Security-Policy".
+func parseCustomResponseHeaders(responseHeaders interface{}, uiHeaders bool) (map[int]map[string][]string, error) {
+	h := make(map[int]map[string][]string)
+	// if responseHeaders is nil, we still should set the default custom headers
+	if responseHeaders == nil {
+		h[0] = map[string][]string{
+			"Strict-Transport-Security": {strictTransportSecurity},
```
Comment (jefferai) [inline listenerutil/parse.go:433]: Please make these header name values const.

File: listenerutil/parse.go [hunk 18]
```diff
@@ -93,6 +94,12 @@ type ListenerConfig struct {
 	CorsAllowedOrigins                       []string    `hcl:"cors_allowed_origins"`
 	CorsAllowedHeaders                       []string    `hcl:"-"`
 	CorsAllowedHeadersRaw                    []string    `hcl:"cors_allowed_headers"`
+
+	// Custom Http response headers
+	CustomApiResponseHeaders    map[int]map[string][]string `hcl:"-"`
```
Comment (jefferai) [inline listenerutil/parse.go:99]: There is an overall issue in the file which is that header names are case-insensitive but the code here won't treat it that way. Go handles this properly if you use `http.Header`. I think reading them in from HCL as strings is probably fine but once you are mixing and matching values below (which I think actually should not be done) then probably actions should take place on an http.Header.

File: listenerutil/parse.go [hunk 19]
```diff
@@ -387,3 +413,142 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
...
+			if err != nil {
+				return nil, err
+			}
+
+			h[status] = headerVal
+		}
+	}
+
+	// setting default headers
+	if h[0] == nil {
```
Comment (jefferai) [inline listenerutil/parse.go:476]: Up above if no values are set then you are setting defaults. At this point here you know that values have been set, but you are adding headers in if they were not set by the user anyway. This would make it impossible for users to _unset_ values we are setting by default that they don't want. I think this entire block probably should be removed.

File: listenerutil/parse.go [hunk 20]
```diff
@@ -387,3 +413,142 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
...
+		}
+		if uiHeaders {
+			h[0]["Content-Security-Policy"] = []string{uiContentSecurityPolicy}
+		} else {
+			h[0]["Content-Security-Policy"] = []string{apiContentSecurityPolicy}
+		}
+		return h, nil
+	}
+
+	customResponseHeader, ok := responseHeaders.([]map[string]interface{})
```
Comment (jefferai) [inline listenerutil/parse.go:445]: Rather than doing these checks here for `[]map[string]interface{}` consider using `mapstructure` to do the decoding for you. Not that this is wrong, but it saves some boilerplate. Probably whether it's worth it depends on if we'd want to allow weak decoding that mapstructure supports so you can e.g. type either `200` or `"200"`. Pros and cons to that...

File: listenerutil/parse.go (hunk 18 above)
Comment (DanielleMiu) [reply listenerutil/parse.go:99]: so this is an interesting one. I thought the same, but it seems somewhere (internal to the hcl parser?) the header names *are* normalized. you can see this happening in the test. `cache-control` values overwrites the `Cache-Control` value, and the `test` values are normalized to `Test`

Comment (jefferai) [reply listenerutil/parse.go:99]: Ah, I didn't look at the function below -- it's happening in `parseHeaders` -- the `textproto.CanonicalMIMEHeaderKey` call. This is the same call that `http.Header` uses internally (at least for now).

It might be nice for `parseHeaders` to return an `http.Header` which is then what is sent around for further use by the system. So `Custom(Ui)ResponseHeaders` becomes `map[int]http.Header`. Thoughts?

File: listenerutil/parse.go (hunk 19 above)
Comment (DanielleMiu) [reply listenerutil/parse.go:476]: so I've added what I think is a happy medium. the defaults are still set if a user doesn't specify that specific header, but if the user specifies an empty header, it'll clear the defaults. this will also be reflected in the config test

File: listenerutil/parse.go (hunk 20 above)
Comment (DanielleMiu) [reply listenerutil/parse.go:445]: would there be any benefits to the hcl by implementing `mapstructure`? I would personally prefer to avoid using an external package unless there are explicit benefits to using it

File: listenerutil/parse.go (hunk 18 above)
Comment (DanielleMiu) [reply listenerutil/parse.go:99]: I think this would be the best way to do it if we had access to the internal workings of ResponseWriter, but the interface we're given only accepts headers in a string/string format, so it would be a bit harder to use. with the map we can just iterate over the key/value pairs

File: listenerutil/parse.go (hunk 19 above)
Comment (jefferai) [reply listenerutil/parse.go]: That makes sense to me

File: listenerutil/parse.go (hunk 20 above)
Comment (jefferai) [reply listenerutil/parse.go:452]: As I said, depends on if we want to allow weak decoding, and also handles a lot of type checking/reflect stuff for us...makes it easy to declare the structure you want and have it populated for you. `mapstructure` is used all over HC code (it's our lib) so if it's helpful it's a totally fine dep to bring in.

File: listenerutil/parse.go (hunk 18 above)
Comment (DanielleMiu) [reply listenerutil/parse.go:99]: if we can still iterate over it then that would work fine. I just meant that the ResponseWriter interface just gives us access to a single http.Header via ResponseWriter.Header() so we need to use Header.Add or Header.Set no matter what.

File: listenerutil/parse.go (hunk 20 above)
Comment (DanielleMiu) [reply listenerutil/parse.go:452]: so I've been reading through the [mapstructure docs](https://pkg.go.dev/github.com/mitchellh/mapstructure), and it seems really helpful for decoding known structures, but I'm not sure how it would help here, since we need to do additional parsing on all the values anyway via `convertStatusCode` and `parseHeaders`

**Review by johanbrandhorst — APPROVED** [review]
LGTM

File: listenerutil/parse.go [hunk 21]
```diff
@@ -387,3 +414,157 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
...
+			headerVal, err := parseHeaders(headerValMap)
+			if err != nil {
+				return nil, err
+			}
+
+			h[status] = headerVal
+		}
+	}
+
+	// setting default headers
```
Comment (sgmiller) [inline listenerutil/parse.go:483]: This could be a bit more generic, a for loop over pairs of header names/defaults for example, but not that important.

**Review by sgmiller — COMMENTED** [review]
My only meta comment is that it seems difficult to prevent the default headers from being added if that was actually desired.

File: listenerutil/parse.go (hunk 21 above)
Comment (DanielleMiu) [reply listenerutil/parse.go:483]: this is done in steps so that more specific status codes can overwrite less specific ones. ex, you can set a header on the default, but the same header on 200, and the 200 header will be returned instead of the default

and regarding preventing defaults, this is also possible by setting empty headers in your config file: https://github.com/hashicorp/go-secure-stdlib/pull/49/files#diff-462be8440ab1a6db8af3debb42b464e02433a564223b166d96dfbfb5702fdfb7R57-R58

File: listenerutil/handler.go (hunk 9 above)
Comment (jefferai) [reply listenerutil/handler.go:72]: I also have this concern. Seems like you might want to say "all 2xx level status codes need header X but 201 also needs header Y".

You _could_ duplicate the headers in the config file between 2xx and 201, but I'm not sure why the current behavior is necessary for overrides to work.

File: listenerutil/listener.go [hunk 22]
```diff
@@ -85,7 +85,7 @@ func TLSConfig(
 	if err := cg.Reload(); err != nil {
 		// We try the key without a passphrase first and if we get an incorrect
 		// passphrase response, try again after prompting for a passphrase
-		if errors.As(err, &x509.IncorrectPasswordError) {
+		if errors.Is(err, x509.IncorrectPasswordError) {
```
Comment (jefferai) [inline listenerutil/listener.go:88]: What's the rationale for this change?

File: listenerutil/parse.go (hunk 20 above)
Comment (jefferai) [reply listenerutil/parse.go:453]: Parsing is separate from validation. Using mapstructure would allow people to specify status codes as either strings or ints. Right now you're forcing headers to be specified as strings, which may not be intuitive since HCL supports ints. I'm fine with it in this case because we also need to support 2xx style values, but just an FYI.
